File: leetcode/python/0202_happy_number.py
class Solution(object):
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        def helper(num):
            res = 0
            while num > 0:
                res += pow(num % 10, 2)
                num /= 10
            return res

        if n == 1: return True

        slow = n
        fast = helper(n)

        while n != 1 and slow != fast:
            slow = helper(slow)
            fast = helper(helper(fast))
            
        return fast == 1

File: leetcode/java/0202_happy_number.java
class Solution {
    public int getNext(int n) {
        int res = 0;
        while (n > 0) {
            res += (n % 10) * (n % 10);
            n /= 10;
        }

        return res;
    }

    public boolean isHappy(int n) {
        int slowRunner = n;
        int fastRunner = getNext(n);

        while (fastRunner != 1 && slowRunner != fastRunner) {
            slowRunner = getNext(slowRunner);
            fastRunner = getNext(getNext(fastRunner));
            System.out.println(slowRunner + "  " + fastRunner);
        }

        return fastRunner == 1;
    }
}